torch/ao/quantization/fx/lstm_utils.py

```diff
@@ -0,0 +1,160 @@
+        cell_qm.set_module_name("output_gate", make_qconfig(sigmoid_obs_ctr))
+    if tanh_obs_ctr is not None:
+        cell_qm.set_module_name("cell_gate", make_qconfig(tanh_obs_ctr))
+
+    # Insert observers into each LSTM cell
+    # TODO: maybe make this work for layer_bw as well
+    for layer in quantizable_lstm.layers:
+        cell = layer.layer_fw.cell
+        cell = prepare_fx(cell, cell_qm, example_inputs, backend_config=backend_config)
+        # HACK: Manually replace the activation_post_process following these ops.
+        # This is needed for FloatFunctional ops because there is currently no way
+        # to configure these ops in FX graph mode quantization today. This is because
+        # the FloatFunctional modules simply disappear from the graph after tracing.
+        # In the future, we should rewrite quantizable LSTM without FloatFunctionals.
+        node_name_to_activation_post_process_ctr = { 
+            "add": linear_output_obs_ctr,  # gates.add
+            "mul": cell_state_obs_ctr,  # fgate_cx.mul
+            "mul_1": cell_state_obs_ctr,  # igate_cgate.mul
+            "add_1": cell_state_obs_ctr,  # fgate_cx_igate_cgate.add
+            "mul_2": hidden_state_obs_ctr,  # ogate_cy.mul
+        }
+        for node in cell.graph.nodes:
+            if node.name not in node_name_to_activation_post_process_ctr:
+                continue
+            assert len(node.users) == 1
+            activation_post_process_name = list(node.users.keys())[0].name
+            activation_post_process_ctr = node_name_to_activation_post_process_ctr[node.name]
+            if activation_post_process_ctr is not None:
+                setattr(cell, activation_post_process_name, activation_post_process_ctr())
+        layer.layer_fw.cell = cell
+    return quantizable_lstm
+
+def _get_reference_quantized_lstm_module(
+    observed_lstm: torch.ao.nn.quantizable.LSTM,
+    backend_config: Optional[BackendConfig] = None,
+) -> torch.ao.nn.quantized.LSTM:
+    """
+    Return a `torch.ao.nn.quantized.LSTM` created from a `torch.ao.nn.quantizable.LSTM`
+    with observers or fake quantizes inserted through `prepare_fx`, e.g. from
+    `_get_lstm_with_individually_observed_parts`.
+
+    This is meant to be used to convert an observed module to a quantized module in the
+    custom module flow.
+
+    Args:
+        `observed_lstm`: a `torch.ao.nn.quantizable.LSTM` observed through `prepare_fx`
+        `backend_config`: BackendConfig to use to produce the reference quantized model
+
+    Return:
+        A reference `torch.ao.nn.quantized.LSTM` module.
+    """
+    for layer in observed_lstm.layers:
+        cell = layer.layer_fw.cell
+        cell = convert_to_reference_fx(cell, backend_config=backend_config)
+        # HACK: Manually remove input quantize nodes and output dequantize nodes,
+        # since custom modules expect quint8 inputs and outputs for now. Note that
```

Comment:
what does the graph look like if we don't do this? For ex, will we have 2 quantize nodes before the linear op? 
Also, confirming that this means the dtype at input/output boundary to the LSTM module will always be fp32 since we are dealing with reference quantized ops.

Reply:
Yes, that was something I spent a long time debugging actually. The problem was because it's a custom module, the caller of this module would feed it quint8 tensors. However, within the module FX graph mode quantization assumes the inputs are fp32, so it'll try to quantize them again, leading to the "Unable to run X with args from QuantizedCPU backend" error. Unfortunately there is no easy way to fix this right now without significant changes to how FX graph mode quantization works.